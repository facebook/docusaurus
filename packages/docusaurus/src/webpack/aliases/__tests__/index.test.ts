/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import {
  loadThemeAliases,
  loadDocusaurusAliases,
  sortAliases,
  createAliasesForTheme,
} from '../index';
import type {LoadedPlugin} from '@docusaurus/types';

describe('sortAliases', () => {
  // https://github.com/facebook/docusaurus/issues/6878
  // Not sure if the risk actually happens, but still made tests to ensure that
  // behavior is consistent
  it('sorts reliably', () => {
    expect(
      Object.values(
        sortAliases({
          '@a/b': 'b',
          '@a/b/c': 'c',
          '@a/b/c/d': 'd',
        }),
      ),
    ).toEqual(['d', 'c', 'b']);
    expect(
      Object.values(
        sortAliases({
          '@a/b': 'b',
          '@a/b/c/d': 'd',
          '@a/b/c': 'c',
        }),
      ),
    ).toEqual(['d', 'c', 'b']);
    expect(
      Object.values(
        sortAliases({
          '@a/b/c/d': 'd',
          '@a/b/c': 'c',
          '@a/b': 'b',
        }),
      ),
    ).toEqual(['d', 'c', 'b']);
    expect(
      Object.values(
        sortAliases({
          '@a/b/c': 'c',
          '@a/b': 'b',
          '@a/b/c/d': 'd',
        }),
      ),
    ).toEqual(['d', 'c', 'b']);
  });
});

describe('createAliasesForTheme', () => {
  it('creates aliases for themePath 1 with components', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const themePath = path.join(fixtures, 'theme-1');
    const alias = await createAliasesForTheme(themePath, true);
    // Testing entries, because order matters!
    expect(Object.entries(alias)).toEqual(
      Object.entries({
        '@theme-original/Footer': path.join(themePath, 'Footer/index.js'),
        '@theme-original/Layout': path.join(themePath, 'Layout.js'),
        '@theme/Footer': path.join(themePath, 'Footer/index.js'),
        '@theme/Layout': path.join(themePath, 'Layout.js'),
      }),
    );
    expect(alias).not.toEqual({});
  });

  it('creates aliases for themePath 1 with components without original', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const themePath = path.join(fixtures, 'theme-1');
    const alias = await createAliasesForTheme(themePath, false);
    // Testing entries, because order matters!
    expect(Object.entries(alias)).toEqual(
      Object.entries({
        '@theme/Footer': path.join(themePath, 'Footer/index.js'),
        '@theme/Layout': path.join(themePath, 'Layout.js'),
      }),
    );
    expect(alias).not.toEqual({});
  });

  it('creates aliases for themePath 2 with components', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const themePath = path.join(fixtures, 'theme-2');
    const alias = await createAliasesForTheme(themePath, true);
    // Testing entries, because order matters!
    expect(Object.entries(alias)).toEqual(
      Object.entries({
        '@theme-original/Layout': path.join(themePath, 'Layout/index.js'),
        '@theme-original/Navbar': path.join(themePath, 'Navbar.js'),
        '@theme-original/NavbarItem/NestedNavbarItem': path.join(
          themePath,
          'NavbarItem/NestedNavbarItem/index.js',
        ),
        '@theme-original/NavbarItem/SiblingNavbarItem': path.join(
          themePath,
          'NavbarItem/SiblingNavbarItem.js',
        ),
        '@theme-original/NavbarItem/zzz': path.join(
          themePath,
          'NavbarItem/zzz.js',
        ),
        '@theme-original/NavbarItem': path.join(
          themePath,
          'NavbarItem/index.js',
        ),

        '@theme/Layout': path.join(themePath, 'Layout/index.js'),
        '@theme/Navbar': path.join(themePath, 'Navbar.js'),
        '@theme/NavbarItem/NestedNavbarItem': path.join(
          themePath,
          'NavbarItem/NestedNavbarItem/index.js',
        ),
        '@theme/NavbarItem/SiblingNavbarItem': path.join(
          themePath,
          'NavbarItem/SiblingNavbarItem.js',
        ),
        '@theme/NavbarItem/zzz': path.join(themePath, 'NavbarItem/zzz.js'),
        '@theme/NavbarItem': path.join(themePath, 'NavbarItem/index.js'),
      }),
    );
    expect(alias).not.toEqual({});
  });

  it('creates aliases for themePath 2 with components without original', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const themePath = path.join(fixtures, 'theme-2');
    const alias = await createAliasesForTheme(themePath, false);
    // Testing entries, because order matters!
    expect(Object.entries(alias)).toEqual(
      Object.entries({
        '@theme/Layout': path.join(themePath, 'Layout/index.js'),
        '@theme/Navbar': path.join(themePath, 'Navbar.js'),
        '@theme/NavbarItem/NestedNavbarItem': path.join(
          themePath,
          'NavbarItem/NestedNavbarItem/index.js',
        ),
        '@theme/NavbarItem/SiblingNavbarItem': path.join(
          themePath,
          'NavbarItem/SiblingNavbarItem.js',
        ),
        '@theme/NavbarItem/zzz': path.join(themePath, 'NavbarItem/zzz.js'),
        '@theme/NavbarItem': path.join(themePath, 'NavbarItem/index.js'),
      }),
    );
    expect(alias).not.toEqual({});
  });

  it('creates themePath with no components', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const themePath = path.join(fixtures, 'empty-theme');
    await fs.ensureDir(themePath);
    const alias = await createAliasesForTheme(themePath, true);
    expect(alias).toEqual({});
  });

  it('creates themePath with no components without original', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const themePath = path.join(fixtures, 'empty-theme');
    await fs.ensureDir(themePath);
    const alias = await createAliasesForTheme(themePath, false);
    expect(alias).toEqual({});
  });

  it('creates nothing for invalid themePath that does not exist', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const themePath = path.join(fixtures, '__noExist__');
    const alias = await createAliasesForTheme(themePath, true);
    expect(alias).toEqual({});
  });
});

describe('getDocusaurusAliases', () => {
  it('returns appropriate webpack aliases', async () => {
    await expect(loadDocusaurusAliases()).resolves.toMatchSnapshot();
  });
});

describe('loadThemeAliases', () => {
  it('next alias can override the previous alias', async () => {
    const fixtures = path.join(__dirname, '__fixtures__');
    const theme1Path = path.join(fixtures, 'theme-1');
    const theme2Path = path.join(fixtures, 'theme-2');

    const alias = await loadThemeAliases({
      siteDir: fixtures,
      plugins: [
        {getThemePath: () => theme1Path},
        {getThemePath: () => theme2Path},
      ] as LoadedPlugin[],
    });

    // Testing entries, because order matters!
    expect(Object.entries(alias)).toMatchSnapshot();
  });
});
