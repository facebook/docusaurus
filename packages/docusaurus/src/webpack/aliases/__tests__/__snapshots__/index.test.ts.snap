// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDocusaurusAliases returns appropriate webpack aliases 1`] = `
{
  "@docusaurus/BrowserOnly": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/BrowserOnly.tsx",
  "@docusaurus/ComponentCreator": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/ComponentCreator.tsx",
  "@docusaurus/ErrorBoundary": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/ErrorBoundary.tsx",
  "@docusaurus/ExecutionEnvironment": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/ExecutionEnvironment.ts",
  "@docusaurus/Head": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/Head.tsx",
  "@docusaurus/Interpolate": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/Interpolate.tsx",
  "@docusaurus/Link": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/Link.tsx",
  "@docusaurus/Noop": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/Noop.ts",
  "@docusaurus/Translate": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/Translate.tsx",
  "@docusaurus/constants": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/constants.ts",
  "@docusaurus/isInternalUrl": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/isInternalUrl.ts",
  "@docusaurus/renderRoutes": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/renderRoutes.ts",
  "@docusaurus/router": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/router.ts",
  "@docusaurus/useBaseUrl": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/useBaseUrl.ts",
  "@docusaurus/useBrokenLinks": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/useBrokenLinks.ts",
  "@docusaurus/useDocusaurusContext": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/useDocusaurusContext.ts",
  "@docusaurus/useGlobalData": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/useGlobalData.ts",
  "@docusaurus/useIsBrowser": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/useIsBrowser.ts",
  "@docusaurus/useIsomorphicLayoutEffect": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/useIsomorphicLayoutEffect.tsx",
  "@docusaurus/useRouteContext": "<PROJECT_ROOT>/packages/docusaurus/src/client/exports/useRouteContext.tsx",
}
`;

exports[`loadThemeAliases next alias can override the previous alias 1`] = `
[
  [
    "@theme-init/Layout",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/Layout/index.tsx",
  ],
  [
    "@theme-original/Error",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/Error/index.tsx",
  ],
  [
    "@theme-original/Footer",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-1/Footer/index.js",
  ],
  [
    "@theme-original/Layout",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/Layout/index.js",
  ],
  [
    "@theme-original/Loading",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/Loading/index.tsx",
  ],
  [
    "@theme-original/Navbar",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/Navbar.js",
  ],
  [
    "@theme-original/NavbarItem/NestedNavbarItem",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/NavbarItem/NestedNavbarItem/index.js",
  ],
  [
    "@theme-original/NavbarItem/SiblingNavbarItem",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/NavbarItem/SiblingNavbarItem.js",
  ],
  [
    "@theme-original/NavbarItem/zzz",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/NavbarItem/zzz.js",
  ],
  [
    "@theme-original/NavbarItem",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/NavbarItem/index.js",
  ],
  [
    "@theme-original/NotFound",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/NotFound/index.tsx",
  ],
  [
    "@theme-original/Root",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/Root/index.tsx",
  ],
  [
    "@theme-original/SiteMetadata",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/SiteMetadata/index.tsx",
  ],
  [
    "@theme/Error",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/Error/index.tsx",
  ],
  [
    "@theme/Footer",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-1/Footer/index.js",
  ],
  [
    "@theme/Layout",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/Layout/index.js",
  ],
  [
    "@theme/Loading",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/Loading/index.tsx",
  ],
  [
    "@theme/Navbar",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/Navbar.js",
  ],
  [
    "@theme/NavbarItem/NestedNavbarItem",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/NavbarItem/NestedNavbarItem/index.js",
  ],
  [
    "@theme/NavbarItem/SiblingNavbarItem",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/NavbarItem/SiblingNavbarItem.js",
  ],
  [
    "@theme/NavbarItem/zzz",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/NavbarItem/zzz.js",
  ],
  [
    "@theme/NavbarItem",
    "<PROJECT_ROOT>/packages/docusaurus/src/webpack/aliases/__tests__/__fixtures__/theme-2/NavbarItem/index.js",
  ],
  [
    "@theme/NotFound",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/NotFound/index.tsx",
  ],
  [
    "@theme/Root",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/Root/index.tsx",
  ],
  [
    "@theme/SiteMetadata",
    "<PROJECT_ROOT>/packages/docusaurus/src/client/theme-fallback/SiteMetadata/index.tsx",
  ],
]
`;
